src/services/messagesService.js

```diff
@@ -268,8 +285,44 @@ class MessagesService {
 
     await this.sendMessage(message);
   }
+
+  saveMessagesToLocalStorage() {
+    const messages = Object.values(store.getState().messages.entities);
+
+    encryptionService
+      .encrypteDataForLocalStore(messages)
+      .then((encryptedData) => localforage.setItem("messages", encryptedData));
+
+    return undefined;
+  }
+
+  async restoreMessagesFromLocalStorage() {
+    const messagesCipheratext = await localforage.getItem("messages");
+
+    const decryptedMessages = await encryptionService.decryptDataFromLocalStore(
+      messagesCipheratext
+    );
+
+    const conversations = Object.values(
+      decryptedMessages.reduce((acc, message) => {
+        if (!message.cid) {
+          return acc;
+        }
+        if (!acc[message.cid]) {
+          acc[message.cid] = { _id: message.cid, messagesIds: [] };
+        }
+        acc[message.cid].messagesIds.push(message._id);
+        return acc;
+      }, {})
+    );
+
+    store.dispatch(addMessages(decryptedMessages));
+    store.dispatch(insertChats(conversations));
+  }
 }
 
 const messagesService = new MessagesService();
 
+window.onbeforeunload = messagesService.saveMessagesToLocalStorage;
```

Comment:
will this event work if user kills the browser?

Reply:
yes